src/bytes_mut.rs

```diff
@@ -636,11 +663,18 @@ impl BytesMut {
                 // sure that the vector has enough capacity.
                 let v = &mut (*shared).vec;
 
-                if v.capacity() >= new_cap {
-                    // The capacity is sufficient, reclaim the buffer
-                    let ptr = v.as_mut_ptr();
+                let v_capacity = v.capacity();
+                let ptr = v.as_mut_ptr();
+                let offset = self.ptr.as_ptr().offset_from(ptr) as usize;
```

Comment:
You could use:

```suggestion
                let offset = v_capacity - self.cap;
```

to calculate the offset here.